transports/webrtc/src/udp_mux.rs

```diff
@@ -0,0 +1,516 @@
+// Copyright 2022 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use async_trait::async_trait;
+use futures::StreamExt;
+use stun::{
+    attributes::ATTR_USERNAME,
+    message::{is_message as is_stun_message, Message as STUNMessage},
+};
+use tokio::{io::ReadBuf, net::UdpSocket};
+use tokio_crate as tokio;
+use webrtc::ice::udp_mux::{UDPMux, UDPMuxConn, UDPMuxConnParams, UDPMuxWriter};
+use webrtc::util::{Conn, Error};
+
+use crate::req_res_chan;
+use futures::channel::oneshot;
+use futures::future::{BoxFuture, FutureExt, OptionFuture};
+use futures::stream::FuturesUnordered;
+use std::{
+    collections::{HashMap, HashSet},
+    io::ErrorKind,
+    net::SocketAddr,
+    sync::Arc,
+    task::{Context, Poll},
+};
+
+const RECEIVE_MTU: usize = 8192;
+
+/// A previously unseen address of a remote who've sent us an ICE binding request.
+#[derive(Debug)]
+pub struct NewAddr {
+    pub addr: SocketAddr,
+    pub ufrag: String,
+}
+
+/// An event emitted by [`UDPMuxNewAddr`] when it's polled.
+#[derive(Debug)]
+pub enum UDPMuxEvent {
+    /// Connection error. UDP mux should be stopped.
+    Error(std::io::Error),
+    /// Got a [`NewAddr`] from the socket.
+    NewAddr(NewAddr),
+}
+
+/// A modified version of [`webrtc::ice::udp_mux::UDPMuxDefault`].
+///
+/// - It has been rewritten to work without locks and channels instead.
+/// - It reports previously unseen addresses instead of ignoring them.
+pub struct UDPMuxNewAddr {
+    udp_sock: UdpSocket,
+
+    /// Maps from ufrag to the underlying connection.
+    conns: HashMap<String, UDPMuxConn>,
```

Comment:
What do folks think of indexing by ufrag AND IP + port?

It might be overly careful, though would prevent an attacker to be able to override the `UDPMuxConn` of another peer by somehow choosing the same ufrag.

Corresponding discussion on the specification: https://github.com/libp2p/specs/pull/412#discussion_r977888406

Reply:
see https://github.com/libp2p/rust-libp2p/pull/2622/commits/226f2930daf9a2614c216546c6084de646a9f26d and https://github.com/libp2p/rust-libp2p/pull/2622#discussion_r982977836